### PR TITLE
Fix line breaks between members when a property has an initializer

### PR DESCRIPTION
**Pull request by DoctorKrolic**
Fixes: https://github.com/dotnet/roslyn/issues/42336

The issue is quite old, so the actual problem changed a bit. Before this PR the output for property with initializer was
```cs
class A
{
  public string Prop { get; } = "xyz";
  public void Foo()
  {
  }
}
```
This is not expected, we want to have an empty line between members.

In this PR I added this missing line break, so the behaviour is the same as in the case without property initializer. Added test cases for both. Put them in the same place, where all other declaration tests currently are

### CODE REVIEW

File: src/Compilers/CSharp/Test/Syntax/Syntax/SyntaxNormalizerTests.cs [hunk 1]
```diff
@@ -1702,6 +1702,28 @@ int p
...
                 """);
+            TestNormalizeDeclaration(
+                "class A{public string Prop{get;}public void Foo(){}}", """
+                class A
+                {
+                  public string Prop { get; }
+
+                  public void Foo()
+                  {
+                  }
```
Comment (Youssef1313) [inline src/Compilers/CSharp/Test/Syntax/Syntax/SyntaxNormalizerTests.cs:1713]: What if there wasn't a method here?

**Comment by Youssef1313** [comment]
> Fixes: https://github.com/dotnet/roslyn/issues/42336

Per your description and issue description, the behavior described in the issue isn't the current behavior. Can you confirm?

So basically the complain in #42336 is already fixed, but this PR wants to handle another scenario.

File: src/Compilers/CSharp/Test/Syntax/Syntax/SyntaxNormalizerTests.cs [hunk 2]
```diff
@@ -1702,6 +1702,28 @@ int p
...
+                }
+                """);
+            TestNormalizeDeclaration(
+                "class A{public string Prop{get;}=\"xyz\";public void Foo(){}}", """
```
Comment (CyrusNajmabadi) [inline src/Compilers/CSharp/Test/Syntax/Syntax/SyntaxNormalizerTests.cs:1717]: raw-string for input as well?  so you don't need to escape the quotes?

**Comment by DoctorKrolic** [comment]
> the behavior described in the issue isn't the current behavior. Can you confirm?

Yes

> So basically the complain in https://github.com/dotnet/roslyn/issues/42336 is already fixed, but this PR wants to handle another scenario.

I mean, I took code example from the issue, and it is still problematic, but in a bit other way. Nevertheless, even if to treat this case as a separate one, the issue should be closed as completed anyway)

**Comment by Youssef1313** [comment]
> I mean, I took code example from the issue, and it is still problematic,

I just wanted to confirm I understand the situation correctly.

File: src/Compilers/CSharp/Test/Syntax/Syntax/SyntaxNormalizerTests.cs (hunk 1 above)
Comment (DoctorKrolic) [reply src/Compilers/CSharp/Test/Syntax/Syntax/SyntaxNormalizerTests.cs:1713]: 100% suspected this question) There are already test cases for this:
https://github.com/dotnet/roslyn/blob/683e22d4a2351e5495e18fbb2f7089f2f1a9679d/src/Compilers/CSharp/Test/Syntax/Syntax/SyntaxNormalizerTests.cs#L1159-L1184
It just very luckely appeared that when the next token is a `}`, execution falls to other branch, where it determines only 1 line break.

**Comment by DoctorKrolic** [comment]
Can I get a review here please?

File: src/Compilers/CSharp/Test/Syntax/Syntax/SyntaxNormalizerTests.cs [hunk 3]
```diff
@@ -1702,6 +1702,28 @@ int p
                   } = 1;
                 }
                 """);
+            TestNormalizeDeclaration(
+                "class A{public string Prop{get;}public void Foo(){}}", """
+                class A
+                {
+                  public string Prop { get; }
```
Comment (CyrusNajmabadi) [inline src/Compilers/CSharp/Test/Syntax/Syntax/SyntaxNormalizerTests.cs:1709]: can you add tests with things like `get;set;`?

File: src/Compilers/CSharp/Portable/Syntax/SyntaxNormalizer.cs [hunk 4]
```diff
@@ -425,6 +425,10 @@ private static int LineBreaksAfterSemicolon(SyntaxToken currentToken, SyntaxToke
             {
                 return 0;
             }
+            else if (currentToken.Parent.IsKind(SyntaxKind.PropertyDeclaration))
+            {
+                return 2;
```
Comment (CyrusNajmabadi) [inline src/Compilers/CSharp/Portable/Syntax/SyntaxNormalizer.cs:430]: i'm a tad bit confused by this.  can you give an example of a property here and what currentToken refers to in it?

Comment (DoctorKrolic) [reply src/Compilers/CSharp/Portable/Syntax/SyntaxNormalizer.cs:430]: Lets take a look at
```cs
class C
{
    public string P { get; set; } = "xyz";
}
```
Here there are 3 `;`s in the property. `;`s of `get` and `set` have a parent of `Get(Set)AccessorDeclaration`. The ending `;` is a direct child of a property. Yes, it was a bit surprising for me as well, but the `;` is not part of `= "xyz"`. You can take close look at parse tree at [sharplab](https://sharplab.io/#v2:EYLgtghglgdgNAFxBAzggPgAQEwAIDCAsAFADeJulumAzNQIwAMuACrqbgOYCmCA3LhS8BAX1wBeXACIAHgE8AXlL4kRQA==).
So the above conditions applies to `;` at the end of a property and adds 2 line breaks there.

Comment (Youssef1313) [reply src/Compilers/CSharp/Portable/Syntax/SyntaxNormalizer.cs:430]: So, are we not consistent between properties with initializers and properties without?

Comment (DoctorKrolic) [reply src/Compilers/CSharp/Portable/Syntax/SyntaxNormalizer.cs:430]: @Youssef1313 Properties without initializers won't even step into this method at that stage. This method is only handling cases after semicolon. If a properry doesn't have an initializer, line breaks are computed around closing brace and that path is already correct

**Comment by CyrusNajmabadi** [comment]
> This is not expected, we want to have an empty line between members.

it's unclear to me that we want this.  this feels like this coudl change a lot of code that currently normalizes.

File: src/Compilers/CSharp/Portable/Syntax/SyntaxNormalizer.cs (hunk 4 above)
Comment (jcouv) [reply src/Compilers/CSharp/Portable/Syntax/SyntaxNormalizer.cs:430]: Consider adding an assertion that this is the initializer's semi-colon. This is a bit stronger than a comment.

File: src/Compilers/CSharp/Test/Syntax/Syntax/SyntaxNormalizerTests.cs [hunk 5]
```diff
@@ -1702,6 +1702,28 @@ int p
...
+
+                  public void Foo()
```
[the remark this answers is not in the document]
Comment (CyrusNajmabadi) [reply src/Compilers/CSharp/Test/Syntax/Syntax/SyntaxNormalizerTests.cs]: No.  Only 'Foo' is banned.  It relates to a global policy checker tool we run which flags particular words based on their problems in some cultures.

File: src/Compilers/CSharp/Portable/Syntax/SyntaxNormalizer.cs (hunk 4 above)
Comment (CyrusNajmabadi) [reply src/Compilers/CSharp/Portable/Syntax/SyntaxNormalizer.cs:436]: > There can be no initializer when the semicolon is attached to the property

I'm not sure what this means.  the semicolon doesn't belong to the initializer, it belongs to the property.  Maybe i'm not quite understanding what you mean here.

**Comment by jcouv** [comment]
> Current single test for all declarations is getting really enormous and hard to navigate in

Agreed. Adding method/test boundaries won't cause that much churn, so should be fine especially if some grouping remains (property initializers, indexers, ...).

File: src/Compilers/CSharp/Portable/Syntax/SyntaxNormalizer.cs [hunk 6]
```diff
@@ -425,6 +425,16 @@ private static int LineBreaksAfterSemicolon(SyntaxToken currentToken, SyntaxToke
             {
                 return 0;
             }
+            // If the current semicolon token belongs to a property
+            // then it is a semicolon at the end of a property typically (but not always) after a property initializer:
+            // public int Prop { get; } = 1; <-- this one
+            // public int Prop { get; }; <-- this produces a syntax error, but the semicolon is still attached to the property
+            // In such cases we need to have 2 line breaks in order to have propper separation between members of a class, struct etc.
+            // Note: case, when the property is the last member and needs only 1 line break after it is handled above (the next token is a closing brace then)
```
Comment (jcouv) [inline src/Compilers/CSharp/Portable/Syntax/SyntaxNormalizer.cs:433]: nit: comment would be better inside the case (just above `return 2;`)

File: src/Compilers/CSharp/Portable/Syntax/SyntaxNormalizer.cs [hunk 7]
```diff
@@ -425,6 +425,16 @@ private static int LineBreaksAfterSemicolon(SyntaxToken currentToken, SyntaxToke
             {
                 return 0;
             }
+            // If the current semicolon token belongs to a property
+            // then it is a semicolon at the end of a property typically (but not always) after a property initializer:
+            // public int Prop { get; } = 1; <-- this one
+            // public int Prop { get; }; <-- this produces a syntax error, but the semicolon is still attached to the property
+            // In such cases we need to have 2 line breaks in order to have propper separation between members of a class, struct etc.
```
Comment (jcouv) [inline src/Compilers/CSharp/Portable/Syntax/SyntaxNormalizer.cs:432]: >propper [](http://example.com/codeflow?start=76&length=7)

typo: proper

File: src/Compilers/CSharp/Portable/Syntax/SyntaxNormalizer.cs (hunk 4 above)
Comment (jcouv) [reply src/Compilers/CSharp/Portable/Syntax/SyntaxNormalizer.cs:436]: nit: The assertion I'm proposing would be along the lines of `((PropertySyntax)currentToken.Parent).Semicolon == currentToken`.

**Review by jcouv — APPROVED** [review]
LGTM Thanks (iteration 6) with nits to consider

**Comment by 333fred** [comment]
@CyrusNajmabadi did you still have concerns about the baseline changes here?

File: src/Compilers/CSharp/Test/Syntax/Syntax/SyntaxNormalizerTests.cs [hunk 8]
```diff
@@ -1703,6 +1703,284 @@ int p
...
+                  public int f;
+                }
+                """);
+            TestNormalizeDeclaration(
+                "class A{public string Prop{get;}public int Prop2{get;set;}}", """
+                class A
+                {
+                  public string Prop { get; }
+
+                  public int Prop2 { get; set; }
```
Comment (CyrusNajmabadi) [inline src/Compilers/CSharp/Test/Syntax/Syntax/SyntaxNormalizerTests.cs:1749]: this test is the thing i'm the most wary about.  do we normally group fields together without newlines between them?  if so, i think we should default to properties being handled the same way.  i'd prefer we resolve this here so we don't have to have multiple PRs changing this.   I could also be convinced this is fine... but i'd like the thoughts of @333fred @jcouv .  To me, even looking at our codebase, props are aggregated together.  So preserving that would be good.

**Review by Youssef1313 — COMMENTED** [review]
Just to note: this can break some unit tests of external source generators that use the API, which is common.

File: src/Compilers/CSharp/Portable/Syntax/SyntaxNormalizer.cs [hunk 9]
```diff
@@ -376,6 +376,16 @@ private static int LineBreaksAfterCloseBrace(SyntaxToken currentToken, SyntaxTok
                 return 0;
             }
 
+            // If we are at the end of a property followed by another property
+            // group them together by having only 1 line break.
+            // The current token here is a closing brace of an accessor list:
+            // public int Prop { get; } <-- this one
+            if (currentTokenParent is AccessorListSyntax &&
+                nextToken is { Parent: PropertyDeclarationSyntax })
```
Comment (Youssef1313) [inline src/Compilers/CSharp/Portable/Syntax/SyntaxNormalizer.cs:384]:
```suggestion
                nextToken.Parent is PropertyDeclarationSyntax)
```

File: src/Compilers/CSharp/Portable/Syntax/SyntaxNormalizer.cs [hunk 10]
```diff
@@ -425,6 +435,19 @@ private static int LineBreaksAfterSemicolon(SyntaxToken currentToken, SyntaxToke
...
+                // If the current semicolon token belongs to a property
+                // then it is a semicolon at the end of a property typically (but not always) after a property initializer:
+                // public int Prop { get; } = 1; <-- this one
+                // public int Prop { get; }; <-- this produces a syntax error, but the semicolon is still attached to the property
+                // In such cases we need to have 2 line breaks in order to have proper separation between members of a class, struct etc.
+                // The only exception is when the next token starts a new property.
+                // In such case we want to group these properties together by having only 1 line break.
+                // Note: case, when the property is the last member and needs only 1 line break after it is handled above (the next token is a closing brace then)
+                Debug.Assert(((PropertyDeclarationSyntax)currentToken.Parent).SemicolonToken == currentToken);
+                return nextToken is { Parent: PropertyDeclarationSyntax } ? 1 : 2;
```
Comment (Youssef1313) [inline src/Compilers/CSharp/Portable/Syntax/SyntaxNormalizer.cs:449]:
```suggestion
                return nextToken.Parent is PropertyDeclarationSyntax ? 1 : 2;
```

**Comment by Youssef1313** [comment]
> @Youssef1313 Well, this is what the CI is for)

I'm referring to *external* source generators that are in different codebases.

Generally, any change to syntax normalizer can affect them, but thinking this one is quite common.

**Comment by 333fred** [comment]
We generally don't treat syntax normalizer's output as a public API. This may be a common change, but a generator author isn't going to observe it until they update their roslyn dependency version in their own tests, and it shouldn't have meaningful impact at generator runtime.

**Comment by 333fred** [comment]
What impact does this PR have on non-auto properties? Will they be grouped in the same way? Consider adding some tests with expression-bodied and full-bodied properties.

File: src/Compilers/CSharp/Portable/Syntax/SyntaxNormalizer.cs [hunk 11]
```diff
@@ -376,6 +376,16 @@ private static int LineBreaksAfterCloseBrace(SyntaxToken currentToken, SyntaxTok
                 return 0;
             }
 
+            // If we are at the end of a property followed by another property
+            // group them together by having only 1 line break.
+            // The current token here is a closing brace of an accessor list:
+            // public int Prop { get; } <-- this one
```
Comment (CyrusNajmabadi) [inline src/Compilers/CSharp/Portable/Syntax/SyntaxNormalizer.cs:382]: note: i would only expect this for props that themselves are only on a single line.  multi-line props would continue to be separated by a blank line.

File: src/Compilers/CSharp/Test/Syntax/Syntax/SyntaxNormalizerTests.cs [hunk 12]
```diff
@@ -1976,6 +2215,43 @@ class A
...
+                """);
+            TestNormalizeDeclaration(
+                """class A{public string Prop{get{}}="xyz";public delegate int D();}""", """
+                class A
+                {
+                  public string Prop
+                  {
+                    get
+                    {
+                    }
```
Comment (Youssef1313) [inline src/Compilers/CSharp/Test/Syntax/Syntax/SyntaxNormalizerTests.cs:2240]: Consider testing things like:

```csharp
public string Prop
{
    get;
    set;
}
```

```csharp
public string Prop
{
    get => string.Empty;
    set _ = value;
}
```

Comment (CyrusNajmabadi) [reply src/Compilers/CSharp/Test/Syntax/Syntax/SyntaxNormalizerTests.cs:2240]: the single line form seems fine to me.  i assume you don't get it if the accessors have bodies?

File: src/Compilers/CSharp/Portable/Syntax/SyntaxNormalizer.cs [hunk 13]
```diff
@@ -425,6 +436,28 @@ private static int LineBreaksAfterSemicolon(SyntaxToken currentToken, SyntaxToke
...
+                // public int Prop { get; } = 1; <-- this one
+                // public int Prop { get; }; <-- this produces a syntax error, but the semicolon is still attached to the property
+                // In such cases we need to have 2 line breaks in order to have proper separation between members of a class, struct etc.
+                // The only exception is when the next token starts a new single-line property.
+                // In such case we want to group these properties together by having only 1 line break.
+                // Note: case, when the property is the last member and needs only 1 line break after it is handled above (the next token is a closing brace then)
+                Debug.Assert(((PropertyDeclarationSyntax)currentToken.Parent).SemicolonToken == currentToken);
+
+                if (IsSingleLineProperty(property) &&
+                    nextToken.Parent is PropertyDeclarationSyntax)
```
Comment (333fred) [inline src/Compilers/CSharp/Portable/Syntax/SyntaxNormalizer.cs:452]: This doesn't seem correct. Please add a test that has a single-line property followed by a manually-implemented property.

**Review by 333fred — APPROVED** [review]
@dotnet/roslyn-compiler for a second review.

**Comment by DoctorKrolic** [comment]
Is this PR forgotten? @jcouv Can you please take a look, since you've already signed this off before. Also probably want a look from @CyrusNajmabadi to verify that all behaviour expectations match

**Review by jcouv — APPROVED** [review]
LGTM Thanks (iteration 15)

**Comment by jcouv** [comment]
@DoctorKrolic Indeed, that had fallen off my radar (we have many open PRs at the moment). Thanks for the ping. Looks good, thanks!

**Comment by jcouv** [comment]
@DoctorKrolic We've had some CI issues. Could you merge latest bits from the `main` branch into this PR? (I do `git fetch origin main` then `git merge origin/main` then push the commit here)

**Comment by DoctorKrolic** [comment]
I preffer to use GitHub UI since I am already in the browser and it doesn't disable auto-merge

**Comment by jcouv** [comment]
Thanks for the contribution @DoctorKrolic!

**Pull request merged**